### ISSUE TITLE
String Literal issue with function overloads 

### ISSUE THREAD

**Issue by folini** [Microsoft/TypeScript#8251]
**TypeScript Version:** 

TypeScript 1.8.30.0

**Issue**

The following code (copy & paste from the [TypeScript documentation](https://www.typescriptlang.org/docs/handbook/advanced-types.html)) doesn't work. It generates a compiler error.

```ts
function createElement(tagName: "img"): HTMLImageElement;
function createElement(tagName: "input"): HTMLInputElement;
function createElement(tagName: string): Element;
// ... more overloads ...

function createElement(tagName: string): Element {
    // ... code goes here ...
}

```

While, just moving the most general signature at the tip of the list will compile correctly.

```ts
function createElement(tagName: string): Element;
function createElement(tagName: "img"): HTMLImageElement;
function createElement(tagName: "input"): HTMLInputElement;
// ... more overloads ...

function createElement(tagName: string): Element {
    // ... code goes here ...
}

```

**Comment by Arnavion** [comment]
What you have in the OP has no error (well, except for a missing `return` in the function body).

The code in the handbook does not have the `tagName: string` overload, which *does* result in `TS2382: Specialized overload signature is not assignable to any non-specialized signature.`. Is that what you meant?

(Using npm 1.8.10 which I believe corresponds to VS 1.8.30.0)

**Comment by yuit** [comment]
The example doesn't produce an error as @Arnavion points out. We made a change to such behavior. The behavior is in the "typescript@next".

**Comment by folini** [comment]
Thanks @Arnavion and @yuit for taking the time to answer my question.

Apparently I changed the code when I copy & pasted removing the issue. 
The issue appears when there is no declaration of the non-specialized signature, and there is only the implementation. This is the original code listed in the documentation that generates an error:

**Code generating the error**
```ts
function createElement(tagName: "img"): HTMLImageElement;
function createElement(tagName: "input"): HTMLInputElement;
// ... more overloads ...

function createElement(tagName: string): Element {
    // ... code goes here ...
    return;
}
```

**Code with no error**
```ts
function createElement(tagName: "img"): HTMLImageElement;
function createElement(tagName: "input"): HTMLInputElement;
function createElement(tagName: string): Element;
// ... more overloads ...

function createElement(tagName: string): Element {
    // ... code goes here ...
    return;
}
```

**Comment by sandersn** [comment]
This was fixed in #6075.
